### PR TITLE
[Java] Initial on device training support

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
Adds support for on device training to the ORT Java API.

All the tests have been ported over from the C# training tests. I think there are some more tests to be written which cover the use of the eval model, and the model export functionality, but we can discuss those and add them before merging this PR.

The integration into the build system will need some changes when #13964 is merged in, it'll need packaging changes to emit two different jars (`onnxruntime-training` and `onnxruntime`), and I've not added the enable training flag to any of the CI pipelines. It looks like there isn't a better way to see if training is available than calling `api->GetTrainingApi` but if it's turned off it unavoidably logs something to stderr. That would be nice to fix as currently all non-training builds of ORT for Java will log that on startup.

There is one missing piece of functionality, which is that the Java API doesn't currently support supplying the output tensors to be written into by the `TrainStep` or `EvalStep` calls. This is because we've not added the equivalent support to `Session.Run` yet, and I want to spend some time designing it properly after this has been merged in as it's a little tricky wrt the memory management.

### Motivation and Context
We'd like to be able to use the on device training from Java.

cc @askhade @baijumeswani

### CODE REVIEW

**Comment by Craigacp** [comment]
This PR has been updated after the training build changes and is now ready for review. The Java training tests are not currently enabled in any of the CI pipelines, I wasn't sure where to add it as the existing training CI pipeline doesn't seem to use ctest to run the tests, though I have added `--build_java` to orttraining-linux-gpu-training-apis to make it be built.

File: java/src/main/java/ai/onnxruntime/OrtSession.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OrtSession.java:518]: >final long nativeHandle; [](http://example.com/codeflow?start=4&length=24)

I suggest introducing getters.

File: java/src/main/native/ai_onnxruntime_OnnxRuntime.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxRuntime.c:28]: >/ required JNI parameters not needed by functions which don't call back into Java. [](http://example.com/codeflow?start=31&length=82)

Suggest commenting out parameter names leaving only type, that would like nicer. JNIEnv* /* jniEnv */,

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:74]: This release section is growing and duplicaterd. Would you not rather use goto: to the specific cleanup at the end?

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:35]: >GetStringChars [](http://example.com/codeflow?start=41&length=14)

Can GetStringChars() fail?

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:3]: This would have been a lot less and safer code in C++

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:708]: > free(outputNames); [](http://example.com/codeflow?start=3&length=19)

Should we continue here?

**Review by yuslepukhin — CHANGES_REQUESTED** [review]
:clock1:

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:35]: Not unless the pointer is null, which I'll add a check for in Java.

File: java/src/main/java/ai/onnxruntime/OrtSession.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OrtSession.java:518]: I don't understand the C# reference? I can make package private get methods in Java, but it didn't seem worth the effort.

File: java/src/main/native/ai_onnxruntime_OnnxRuntime.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OnnxRuntime.c:28]: This is the idiom I've used everywhere else in the JNI code, do you want it changed everywhere?

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:3]: Well it might have been if I'd written any C++ in the past 15 years, but as I haven't I'm not sure it would turn out much better.

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:708]: Good point, it should throw an error message.

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:74]: I think I'll split it into two methods and use the ifdef to dispatch to them, then use gotos in the windows case. I didn't like the combination of goto and ifdef I had when I was developing it as it was a bit unpleasant to reason over.

Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:74]: I refactored it further as NULL is used as a signalling mechanism for the underlying CreateTrainingSession call, and the Windows logic was broken in that case.

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c]: There already is a null check in Java for `trainPath` and I refactored the logic for the other two.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h
Comment (baijumeswani) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:7]: is this needed?

Comment (Craigacp) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:7]: Yeah, for `bool` support in C11 otherwise the type name is `_Bool` - https://en.cppreference.com/w/c/types/boolean

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml
Comment (baijumeswani) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:25]: This change here will only build the training api for java. But will not run the unit tests considering we have `--skip_tests`. Is it possible to run training specific tests (using some keyword filters) from https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/test/python/orttraining_test_ort_apis.py?

We can get this in another PR if this is not straightforward.

**Comment by baijumeswani** [comment]
training specific additions looks good to me. I'll kick off the CI builds for other pipelines.
Deferring to @yuslepukhin for any Java related comments.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml
Comment (Craigacp) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:25]: The Java tests are run by gradle, and need a specific set of command line arguments to enable the right EP tests and to point at the native library (as when the tests are run the jar package hasn't been fully created yet). Plus they need to be run in the right directory so they can reference the appropriate onnx test models which live elsewhere in the source tree outside the `java` folder.

So it's a bit of a pain to do, and I wasn't sure how to make the necessary modifications to that python as I'm not sure what environment it's running in.

Comment (Craigacp) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:25]: I think it might be worth having another meeting to discuss how to integrate the tests, it'll probably be easier than doing it over email.

**Comment by Craigacp** [comment]
I pushed a fix for the Windows failure, I typo'd the name of something in a Windows specific macro. Not sure why the macOS run is failing, maybe it's picking up gradle from a different place than usual? Did something change in the environment definition?

**Comment by Craigacp** [comment]
Ahh, the macOS failure is fixed in #14733. I'll rebase on that once it's been merged in.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-training-apis.yml:25]: Sounds good. We can talk about this over a call. I think we should not block this PR on that though.

File: java/src/main/java/ai/onnxruntime/OrtSession.java
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OrtSession.java:518]: Yes, this is exactly what I have in mind. From the C# reference, we had to re-work a few times where the native handles are stored or how they are obtained. Getters would allow doing it without refactoring. Making private handles public is not good.

File: java/src/main/native/ai_onnxruntime_OnnxRuntime.c
Comment (yuslepukhin) [reply java/src/main/native/ai_onnxruntime_OnnxRuntime.c:28]: It would be nice, less code and more readability. Feel free to address it in a separate PR.

Comment (yuslepukhin) [reply java/src/main/native/ai_onnxruntime_OnnxRuntime.c:28]: At your discretion.

File: java/src/main/java/ai/onnxruntime/OrtSession.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OrtSession.java]: Done. The package private fields are now private, and there are package private accessors.

File: java/src/main/native/ai_onnxruntime_OrtTrainingSession.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:20]: > wcsncpy_s(outputStr, outputStrLength + 1, (const wchar_t*)charArr, strLength); [](http://example.com/codeflow?start=3&length=79)

The source does not have null terminator.
I have studied the docs and the function will not add a null terminator except by accident, that is the extra + 1 character contains null.
So it has to be added manually. Like:
```.c
outputStr[outputStrLength] = 0;
```

Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:20]: That's why the memory is allocated with calloc to init everything to zero. I can move it to malloc and explicitly set the zero byte if you want.

Comment (yuslepukhin) [reply java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:20]: No need. I am forgetting C.

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Review by baijumeswani — DISMISSED** [review]

**Review by jchen351 — APPROVED** [review]

**Comment by baijumeswani** [comment]
Build seems to be failing on the windows ci pipeline. Not sure if you have access to view the error:

```sh
##[error]java\src\main\native\ai_onnxruntime_OrtTrainingSession.c(20,42): Error C2065: 'outputStrLength': undeclared identifier
   154>D:\a\_work\1\s\java\src\main\native\ai_onnxruntime_OrtTrainingSession.c(20,42): error C2065: 'outputStrLength': undeclared identifier [D:\a\_work\1\b\Debug\onnxruntime4j_jni.vcxproj]
       ClCompile:
```

**Comment by Craigacp** [comment]
> Build seems to be failing on the windows ci pipeline. Not sure if you have access to view the error:
> 
> ```shell
> ##[error]java\src\main\native\ai_onnxruntime_OrtTrainingSession.c(20,42): Error C2065: 'outputStrLength': undeclared identifier
>    154>D:\a\_work\1\s\java\src\main\native\ai_onnxruntime_OrtTrainingSession.c(20,42): error C2065: 'outputStrLength': undeclared identifier [D:\a\_work\1\b\Debug\onnxruntime4j_jni.vcxproj]
>        ClCompile:
> ```

Ahh. I'd fixed that and pushed it out but I think it got lost when I force pushed the rebase to pull in the macOS gradle fix in #14733. I've pushed the fix back out again.

**Comment by Craigacp** [comment]
I got the tests up and running locally on Windows 11 with VS 2022, and things compile and tests pass. I'm not sure which VS is used to compile it in the CI, but hopefully there shouldn't be any more Windows errors now.

**Comment by Craigacp** [comment]
That failure looks exactly like this compiler issue which is marked as fixed? https://developercommunity.visualstudio.com/t/C-compiler:-yield-warning-C4090-incorrec/1133408.

It flags lines 721 and 752 where the allocated memory is passed to free. I can cast it to `(void*)` in the free call if that will make it happy, but it feels like the wrong solution.

**Comment by baijumeswani** [comment]
@Craigacp shall we instead suppress the warning C4090 for those two lines?

**Comment by Craigacp** [comment]
> @Craigacp shall we instead suppress the warning C4090 for those two lines?

I would be fine with that, the code compiles cleanly and passes the tests using VS 2022 so I think it's an over-eager static analyzer, but I didn't want to suppress those warnings without checking with you first.

**Comment by baijumeswani** [comment]
yeah, I didn't see anything wrong with those two lines. Probably, we can suppress the warning.

**Comment by yuslepukhin** [comment]
> > @Craigacp shall we instead suppress the warning C4090 for those two lines?
> 
> I would be fine with that, the code compiles cleanly and passes the tests using VS 2022 so I think it's an over-eager static analyzer, but I didn't want to suppress those warnings without checking with you first.

The compiler is correct, the cast would be wrong. What is the reason outputNames were qualified as const if we are going to modify that memory and then free it? The compiler is asking you do you really understand what you are doing?So the question why?

Is it not being mixed up with char* const* outputNames? If it is mixed up, then what is the reason for the latter const would be?

**Comment by Craigacp** [comment]
> > > @Craigacp shall we instead suppress the warning C4090 for those two lines?
> > 
> > 
> > I would be fine with that, the code compiles cleanly and passes the tests using VS 2022 so I think it's an over-eager static analyzer, but I didn't want to suppress those warnings without checking with you first.
> 
> The compiler is correct, the cast would be wrong. What is the reason outputNames were qualified as const if we are going to modify that memory and then free it? The compiler is asking you do you really understand what you are doing? So the question why?
> 
> Is it not being mixed up with `char* const* outputNames`? If it is mixed up, then what is the reason for the latter const would be?
> 
> Typically, when the code receives const type* it is a signal, do not free it, you do not own it.

The same definition which it is complaining about at line 721 in OrtTrainingSession.c appears multiple times in that file, and also in OrtSession.c and it doesn't have a warning in those other locations, nor does this warning occur in VC 2022. The code allocates some memory to use as an array to store `const char *` values which we get out of the JNI function, and I need to free the allocated memory. The `const char *` values are returned to the JVM and it deals with them.

My understanding of the definition `const char **` is that it is a pointer to a pointer of constant `char`s. The JNI function `getStringUTFChars` returns a `const char *`, if I change the definition of `outputNames` to `char ** outputNames` then I get a compile error out of clang as the return value is being assigned to a non-const char pointer: 
```
onnxruntime/java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:727:20: error: assigning to 'char *' from 'const char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
    outputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaOutputStrings[i], NULL);
```

So I'm pretty sure the code is correct, and it compiles fine on macOS, Linux and VC 2022.

**Comment by yuslepukhin** [comment]
> > > > @Craigacp shall we instead suppress the warning C4090 for those two lines?
> > > 
> > > 
> > > I would be fine with that, the code compiles cleanly and passes the tests using VS 2022 so I think it's an over-eager static analyzer, but I didn't want to suppress those warnings without checking with you first.
> > 
> > 
> > The compiler is correct, the cast would be wrong. What is the reason outputNames were qualified as const if we are going to modify that memory and then free it? The compiler is asking you do you really understand what you are doing? So the question why?
> > Is it not being mixed up with `char* const* outputNames`? If it is mixed up, then what is the reason for the latter const would be?
> > Typically, when the code receives const type* it is a signal, do not free it, you do not own it.
> 
> The same definition which it is complaining about at line 721 in OrtTrainingSession.c appears multiple times in that file, and also in OrtSession.c and it doesn't have a warning in those other locations, nor does this warning occur in VC 2022. The code allocates some memory to use as an array to store `const char *` values which we get out of the JNI function, and I need to free the allocated memory. The `const char *` values are returned to the JVM and it deals with them.
> 
> My understanding of the definition `const char **` is that it is a pointer to a pointer of constant `char`s. The JNI function `getStringUTFChars` returns a `const char *`, if I change the definition of `outputNames` to `char ** outputNames` then I get a compile error out of clang as the return value is being assigned to a non-const char pointer:
> 
> ```
> onnxruntime/java/src/main/native/ai_onnxruntime_OrtTrainingSession.c:727:20: error: assigning to 'char *' from 'const char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
>     outputNames[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaOutputStrings[i], NULL);
> ```
> 
> So I'm pretty sure the code is correct, and it compiles fine on macOS, Linux and VC 2022.

Here is my concern. We are inserting casts left and right that affects all compilers and platforms without due diligence.
My recommendation in this case is to suppress the warning only for this compiler and only for this function using pragmas.
An example for this can be found [here](https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/common/inlined_containers_fwd.h#L10).

**Comment by Craigacp** [comment]
Sure, which is why I didn't want to cast it. I'm fine with suppressing the warning for whatever the minimal amount of code is, line preferably but function otherwise. I do think this is an over-aggressive static analyzer given that the same idiom appears in many other places without issue, and so disabling just this warning with a note to remove it when the baseline becomes VC 2022 is my preferred solution, unless there's a different code shape that will stop it from getting confused.

**Comment by yuslepukhin** [comment]
> Sure, which is why I didn't want to cast it. I'm fine with suppressing the warning for whatever the minimal amount of code is, line preferably but function otherwise. I do think this is an over-aggressive static analyzer given that the same idiom appears in many other places without issue, and so disabling just this warning with a note to remove it when the baseline becomes VC 2022 is my preferred solution, unless there's a different code shape that will stop it from getting confused.

I thought it is a compiler. I agree to suppress it on the function level.

**Review by baijumeswani — APPROVED** [review]

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Comment by baijumeswani** [comment]
[Craigacp](https://github.com/Craigacp) Would you please resolve the conflict. I will make sure this is merged by tomorrow. Sorry for all the CI build errors, some pipelines were not working well.

**Comment by Craigacp** [comment]
I rebased the branch.

**Review by baijumeswani — APPROVED** [review]

**Comment by baijumeswani** [comment]
@craigacp will https://github.com/microsoft/onnxruntime/pull/14862 have any impact on this PR?

**Comment by mszhanyi** [comment]
Could you sync with @edgchen1 if there's potential issue with [#144862](https://github.com/microsoft/onnxruntime/pull/14862)

**Comment by Craigacp** [comment]
I think both PRs should merge cleanly. While they do both modify `onnxruntime_java.cmake` and `build.gradle` the modifications are in separate parts of the files and there isn't any interaction.

**Review by mszhanyi — APPROVED** [review]

File: cmake/onnxruntime_java.cmake
Comment (Craigacp) [reply cmake/onnxruntime_java.cmake:1]: That file was modified in 2022, I don't think I've made changes to it since then. Oracle's preferred copyright statement at the moment is `Copyright (c) <year-created>, <year-last-modified>, Oracle and/or its affiliates. All rights reserved.` so I've been updating them in that form.

**Comment by baijumeswani** [comment]
Thank you for the contribution @Craigacp. And thanks for being so patient with the process.

**Comment by baijumeswani** [comment]
@Craigacp, it seems like this PR introduced a failing test in the orttraining mac ci pipeline. Example run: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=909853&view=logs&j=ded01483-6627-58ac-64dc-d4a232827e5d&t=259355ad-c348-581d-89bb-38ce8fc55599

```sh
7: 
7: TensorCreationTest > testScalarCreation() PASSED
7: 
7: TrainingTest > TestTrainingSessionTrainStep() SKIPPED
7: libc++abi: terminating with uncaught exception of type std::out_of_range: vector
7: 
7: TrainingTest > TestTrainingSessionOptimizerStep() SKIPPED
```

Did you observe a similar error on your end when you ran the tests?

**Comment by yuslepukhin** [comment]
> I saw that once or twice when I was developing it, but it didn't consistently replicate and it didn't fail during the test runs I was doing in the past few weeks on macOS. I don't use C++ vectors in the JNI C code I wrote, and I'd expect all the ORT C API methods to catch the exceptions and report error codes so I'm not sure where it's coming from. It also doesn't appear to create a JVM log file which is what usually happens when something brings down the JVM in native code.
> 
> I thought all the tests passed in the CI before the PR was merged in?
> 
> The link you gave has it happening in a slightly different place, which is more worrying if it's moving around.
> 
> ```
> 7: 
> 7: TrainingTest > TestTrainingSessionTrainStep() SKIPPED
> 7: 
> 7: TrainingTest > TestTrainingSessionOptimizerStep() PASSED
> 7: 2023-03-08 20:09:13.930 java[92351:439149] 2023-03-08 20:09:13.930149 [W:onnxruntime:ort-java, checkpoint.cc:311 OrtSaveInternal] Checkpoint directory exists - data may be overwritten.
> 7: libc++abi: terminating with uncaught exception of type std::out_of_range: vector
> 7: 
> 7: TrainingTest > TestSaveCheckpoint() SKIPPED
> 7: 
> ```
> 
> I'll try to replicate it across the different machines I have access to tomorrow and see if I can get a stack trace out of it.

I just saw it yesterday, it is a bug. I notified training team.

**Comment by Craigacp** [comment]
> > I saw that once or twice when I was developing it, but it didn't consistently replicate and it didn't fail during the test runs I was doing in the past few weeks on macOS. I don't use C++ vectors in the JNI C code I wrote, and I'd expect all the ORT C API methods to catch the exceptions and report error codes so I'm not sure where it's coming from. It also doesn't appear to create a JVM log file which is what usually happens when something brings down the JVM in native code.
> > I thought all the tests passed in the CI before the PR was merged in?
> > The link you gave has it happening in a slightly different place, which is more worrying if it's moving around.
> > ```
> > 7: 
> > 7: TrainingTest > TestTrainingSessionTrainStep() SKIPPED
> > 7: 
> > 7: TrainingTest > TestTrainingSessionOptimizerStep() PASSED
> > 7: 2023-03-08 20:09:13.930 java[92351:439149] 2023-03-08 20:09:13.930149 [W:onnxruntime:ort-java, checkpoint.cc:311 OrtSaveInternal] Checkpoint directory exists - data may be overwritten.
> > 7: libc++abi: terminating with uncaught exception of type std::out_of_range: vector
> > 7: 
> > 7: TrainingTest > TestSaveCheckpoint() SKIPPED
> > 7: 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'll try to replicate it across the different machines I have access to tomorrow and see if I can get a stack trace out of it.
> 
> I just saw it yesterday, it is a bug. I notified training team.

Yeah, I've run it down now, I think it's because in Java I call the methods to ask how many inputs and outputs the eval model has all the time, even if there isn't an eval model, and the accessors don't catch exceptions which bubble up and kill the JVM.

**Comment by baijumeswani** [comment]
Asking how many inputs/outputs should return 0 even though the eval graph is not present.

Are you also querying the input/output names at any index even though the eval graph is not provided?

**Comment by baijumeswani** [comment]
It might be possible that the query for eval model inputs/outputs is returning an uninitialized int in case the eval model was not provided. I will try to run the pipeline against a branch which explicitly initializes the variable to see if that will resolve this.

**Pull request merged**
